Fixed #29186 -- Fixed pickling HttpRequest and subclasses.

Anv3sh opened the pull request:
Introduced `DjangoSocketHandler` class to handle socket logging which simply removes all the non-picklable attributes from the request object.

felixxm commented on django/utils/log.py:
```diff
@@ -142,6 +142,41 @@ def format_subject(self, subject):
         return subject.replace("\n", "\\n").replace("\r", "\\r")
 
 
+class DjangoSocketHandler(logging.handlers.SocketHandler):
```
Rather, I was thinking about stripping non-picklable from `HttpRequest` directly in the `HttpRequest.__getstate__()`.

Anv3sh replied:
My bad I thought you could have meant that but didn't follow through 🙇‍♂️

Anv3sh replied:
So I should just add `HttpRequest.__getstate__ ()`  with the non-pickalable frozenset and depricate the `DjangoSocketHandler` class, is that right?

felixxm replied:
There is no need to deprecate a class that never existed in Django. Just remove it from the patch.

felixxm commented:
@Anv3sh Do you have time to keep working on this?

Anv3sh commented:
> @Anv3sh Do you have time to keep working on this?

Yes I'm working on it.
```py
non_picklable_attrs = frozenset(
           [
            "resolver_match",
            "environ",
            "META",
            "_stream",
           ]
   ) 
```
Couple of things that I'm confused with is that if we keep the `environ` and `META` attrs as non-picklable it works fine while socket logging and gets pickled but since `__getstate()__` is used by the `.as_view()` it needs the `environ` and `META` and throws exception in some of the testcases.

The `HttpRequest.__getstate__( )` looks as following :
```py
     def __getstate__(self):
        obj_dict = self.__dict__.copy()
        for attr in self.non_picklable_attrs:
            if attr in obj_dict:
                del obj_dict[attr]
        return obj_dict
```

felixxm commented:
> Couple of things that I'm confused with is that if we keep the `environ` and `META` attrs as non-picklable it works fine while socket logging and gets pickled but since `__getstate()__` is used by the `.as_view()` it needs the `environ` and `META` and throws exception in some of the testcases.

`as_view()` doesn't use `__getstate__()`. Tests errors (e.g. `auth_tests.test_templates.AuthTemplateTests.test_password_reset_view`) are related with creating `request` in the `setUpTestData()` which uses `__dict__` and `deepcopy()` of attributes to ensure isolation.

The first question is: Is the `environ` really non-picklable? If yes, can we try to override `HttpRequest.__deepcopy_()`?

Anv3sh commented:
> > Couple of things that I'm confused with is that if we keep the `environ` and `META` attrs as non-picklable it works fine while socket logging and gets pickled but since `__getstate()__` is used by the `.as_view()` it needs the `environ` and `META` and throws exception in some of the testcases.
> 
> `as_view()` doesn't use `__getstate__()`. Tests errors (e.g. `auth_tests.test_templates.AuthTemplateTests.test_password_reset_view`) are related with creating `request` in the `setUpTestData()` which uses `__dict__` and `deepcopy()` of attributes to ensure isolation.
> 
> The first question is: Is the `environ` really non-picklable? If yes, can we try to override `HttpRequest.__deepcopy_()`?

Got it 👍🏻. As far as I tested the `environ` isn't picklable so as you said we might have to override the `HttpRequest.__deepcopy__()` but I don't have much idea on how we can proceed with that if you could give some suggestions?

Anv3sh commented:
@felixxm overrided the `HttpRequest.__deepcopy__()` still don't know if we should treat `META` as a non-picklable attribute as I ran this custom test: 
```py
def test_pickle_request(self):
        request = HttpRequest()
        request.method = "GET"
        request.COOKIES = {"post-key": "post-value"}
        dump = pickle.dumps(request)
        request_from_pickle = pickle.loads(dump)
        self.assertEqual(repr(request), repr(request_from_pickle))
``` 
I found out that the `HttpResponse.__repr__()` needs the `META.QUERY_STRING` attribute but if we treat `META` as a picklable attribute the socket logging throws error. If you could take a look 🙇‍♂️

felixxm commented:
@Anv3sh Please add tests.

Anv3sh commented:
> @Anv3sh Please add tests.

Can you please review this comment 🙇‍♂️

felixxm commented:
> I found out that the HttpRequest.__repr__() needs the META.QUERY_STRING attribute but if we treat META as a picklable attribute the socket logging throws error. If you could take a look

So, again, Can you add a test that crashes when `META` in not in the `non_picklable_attrs`? then we could discuss what's inside `META`. As far as I'm aware, `META` should generally be treated as non-picklable even though it can be picklable in some basic cases.

Anv3sh commented:
> > I found out that the HttpRequest.**repr**() needs the META.QUERY_STRING attribute but if we treat META as a picklable attribute the socket logging throws error. If you could take a look
> 
> So, again, Can you add a test that crashes when `META` in not in the `non_picklable_attrs`? then we could discuss what's inside `META`. As far as I'm aware, `META` should generally be treated as non-picklable even though it can be picklable in some basic cases.

Ok I will add the test case and remove the `META` attribute from the `non_picklable_attrs` for the test to pass and then we can add it later after checking.

Anv3sh commented:
Added the test for request pickling and sorry for the linting errors missed checking for them.

felixxm commented on django/http/request.py:
```diff
@@ -78,6 +87,20 @@ def __repr__(self):
+
+    def __deepcopy__(self, memo):
+        obj = copy.copy(self)
+        obj.environ = copy.deepcopy(self.environ, memo)
+        obj.META = copy.deepcopy(self.META, memo)
```
We should have a more generic solution and deep-copy all non-picklable attributes, e.g.
```python
        for attr in self.non_picklable_attrs:
            if hasattr(self, attr):
                setattr(obj, attr, copy.deepcopy(getattr(self, attr), memo))
```

felixxm commented on django/http/request.py:
```diff
@@ -51,6 +51,15 @@ class HttpRequest:
+        [
+            "resolver_match",
+            "environ",
+            "META",
```
I'd be good to identify what in `META` is non-pickleable and maybe remove only this value :thinking:

felixxm commented on tests/requests/tests.py:
```diff
@@ -669,6 +670,20 @@ def read(self, len=0):
+        self.assertEqual(
+            str(cm.exception),
+            "'HttpRequest' object has no attribute 'META'",
+        )
```
We should also add tests for actual requests (`WSGIRequest`, `ASGIRequest`) which crash when removing any key from `non_picklable_attrs`.

Anv3sh commented:
Added the `META` attributes that were non_picklable to `META_non_picklable_attrs` and changed the `test_pickling_request` according to that. Still not able to figure out a test for actual requests( `WSGIRequest` ,`ASGIRequest`) which crash when removing any key from `non_picklable_attrs`.

felixxm commented on django/http/request.py:
```diff
@@ -51,6 +51,21 @@ class HttpRequest:
+            "_stream",
+        ]
+    )
+    META_non_picklable_attr = frozenset(
+        [
+            "wsgi.errors",
+            "wsgi.input",
+        ]
+    )
```
`wsgi.errors` and `wsgi.input` are specific for `WSGIRequest` so we should move this to the `WSGIRequest.__getstate__()`. It seems that `META` is picklable for `HttpRequest` and `ASGIRequest`.

Anv3sh replied:
Ok will do it 👍🏻  and yes it seems `META` is picklable for `HttpRequest` and `ASGIRequest`.

Anv3sh replied:
@felixxm so the WSGIRequest.__getstate__() and WSGIRequest.__deepcopy__() would look something like this :
```py
non_picklable_attrs = frozenset(
        [
            "wsgi.errors",
            "wsgi.input",
        ]
    )
    def __getstate__(self):
        obj_dict = self.__dict__.copy()
        for attr in self.non_picklable_attrs:
            if attr in obj_dict["META"]:
                del obj_dict["META"][attr]
        super().__getstate__()

    def __deepcopy__(self, memo):
        obj = copy.copy(self)
        for attr in self.non_picklable_attrs:
            if hasattr(self, attr):
                setattr(obj, attr, copy.deepcopy(getattr(self, attr), memo))
        memo[id(self)] = obj
        super().__deepcopy__(memo)
```

Anv3sh commented:
@felixxm added the `WSGIRequest.__getstate__()` and `WSGIRequest.wsgi_non_picklable_attrs`

felixxm commented on django/core/handlers/wsgi.py:
```diff
@@ -64,6 +64,13 @@ def close(self):
 
 
 class WSGIRequest(HttpRequest):
+    wsgi_non_picklable_attrs = frozenset(
```
I'd rename it to `meta_non_picklable_items`.

felixxm commented on django/core/handlers/wsgi.py:
```diff
@@ -89,6 +96,13 @@ def __init__(self, environ):
         self._read_started = False
         self.resolver_match = None
 
+    def __getstate__(self):
+        obj_dict = self.__dict__.copy()
+        for attr in self.wsgi_non_picklable_attrs:
+            if attr in obj_dict["META"]:
+                del obj_dict["META"][attr]
+        return super().__getstate__()
```
This should be called first, i.e.
```suggestion
    def __getstate__(self):
        state = super().__getstate__()
        for attr in self.meta_non_picklable_items:
            if attr in state["META"]:
                del state["META"][attr]
        return state
```

felixxm commented on django/http/request.py:
```diff
@@ -51,6 +51,14 @@ class HttpRequest:
+    non_picklable_attrs = frozenset(
+        [
+            "resolver_match",
+            "environ",
```
`environ` is `WSGI` specific we should move it to the `WSGIRequest`:
```python
class WSGIRequest(HttpRequest):
    non_picklable_attrs = HttpRequest.non_picklable_attrs | frozenset(["environ"])
    meta_non_picklable_items = frozenset(
        ...
```

felixxm reviewed (commented):
@Anv3sh Thanks :+1: 

This patch should allow for removing `asgi_request` and `wsgi_request` from `HttpResponse.non_picklable_attrs`.

Anv3sh commented:
@felixxm committed the requested changes don't know why one of the checks failed.

felixxm commented:
@Anv3sh Thanks :+1:

We don't need more tests because tests added in d7f5bfd241666c0a76e90208da1e9ef81aec44db will check that `WSGIRequest` and `ASGIRequest` are picklable (after removing `wsgi_request` and `asgi_request` from `HttpResponse.non_picklable_attrs`).

> @felixxm committed the requested changes don't know why one of the checks failed.

Don't worry this test is flaky.

vamsi-verma-s commented:
Hi @felixxm, will this fix be back-ported to older versions? if so, any timeline?

vamsi-verma-s commented:
nvm, just read the [doc](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions)

salomvary commented:
I don't really understand what the purpose of this change was, but it broke code in Django 4.2 that tries to "clone" and override requests using `copy`.

Something like this was working before:

```
request = HttpRequest()
request_copy = copy(request)
match = request_copy.resolver_match

AttributeError: 'WSGIRequest' object has no attribute 'resolver_match'
```

felixxm commented:
@salomvary Thanks for the report.

> Something like this was working before:

It worked only because `resolver_match` was `None`. However, I agree that we could make it less backward compatible, e.g.:
```
diff --git a/django/http/request.py b/django/http/request.py
index 2ef9dfd649..a82851f660 100644
--- a/django/http/request.py
+++ b/django/http/request.py
@@ -87,7 +87,7 @@ class HttpRequest:
     def __getstate__(self):
         obj_dict = self.__dict__.copy()
         for attr in self.non_picklable_attrs:
-            if attr in obj_dict:
+            if obj_dict.get(attr, None) is not None:
                 del obj_dict[attr]
         return obj_dict
 
```
Please create a new ticket in Trac and follow our [bug reporting guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/bugs-and-features/#reporting-bugs). All bugfixes require [an accepted ticket](https://code.djangoproject.com/newticket).

salomvary commented:
Thanks for the quick response!

> It worked only because `resolver_match` was `None`.

I don't see why shallow copying a `HttpRequest` with a non-None `resolver_match` should be a problem. Am I missing something? (I'm not really familiar with the internals of `copy` and `pickle`.)

For me it sounds like this was a change that was necessary for `pickle` to work on `HttpRequest` which inadvertently affected how `copy` now behaves.

felixxm commented:
Some attributes of `HttpRequest` and `HttpResponse` are not picklable (see ticket-32969, ticket-29186, and ticket-23895 for more details and some discussion) which breaks e.g. logging. That's why we decided to strip non-picklable attributes.

Anv3sh commented:
> @salomvary Thanks for the report.
> 
> > Something like this was working before:
> 
> It worked only because `resolver_match` was `None`. However, I agree that we could make it less backward compatible, e.g.:
> 
> ```
> diff --git a/django/http/request.py b/django/http/request.py
> index 2ef9dfd649..a82851f660 100644
> --- a/django/http/request.py
> +++ b/django/http/request.py
> @@ -87,7 +87,7 @@ class HttpRequest:
>      def __getstate__(self):
>          obj_dict = self.__dict__.copy()
>          for attr in self.non_picklable_attrs:
> -            if attr in obj_dict:
> +            if obj_dict.get(attr, None) is not None:
>                  del obj_dict[attr]
>          return obj_dict
>  
> ```
> 
> Please create a new ticket in Trac and follow our [bug reporting guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/bugs-and-features/#reporting-bugs). All bugfixes require [an accepted ticket](https://code.djangoproject.com/newticket).

Should I create the required ticket?? 😄

felixxm commented:
> Should I create the required ticket?? smile

Yes please, I think we should fix cloning an empty `HttpRequest` and `HttpResponse` objects with `None` attributes (regressions in d7f5bfd241666c0a76e90208da1e9ef81aec44db and 6220c445c40a6a7f4d442de8bde2628346153963).

Anv3sh commented:
> > Should I create the required ticket?? smile
> 
> Yes please, I think we should fix cloning an empty `HttpRequest` and `HttpResponse` objects with `None` attributes (regressions in [d7f5bfd](https://github.com/django/django/commit/d7f5bfd241666c0a76e90208da1e9ef81aec44db) and [6220c44](https://github.com/django/django/commit/6220c445c40a6a7f4d442de8bde2628346153963)).

Got it 👍🏻
